javascript/javascript1/week4/voice assistant.js

```diff
@@ -0,0 +1,90 @@
+//Voice Assistant
+let userName = "";
+let todos = [];
+
+function getReply(command) {
+    // (1) Save user name
+    if (command.startsWith("Hello my name is")) {
+        userName = command.replace("Hello my name is", "").trim();
+        return `Nice to meet you, ${userName}`;
+    }
+
+    // (2) Respond with user name
+    else if (command === "What is my name?") {
+        return userName ? `Your name is ${userName}` : "I don't know your name yet!";
+    }
+
+    // (3) Add an item to the todo list
+    else if (command.startsWith("Add")) {
+        let todoItem = command.replace("Add", "").replace("to my todo", "").trim();
+        todos.push(todoItem);
+        return `${todoItem} added to your todo.`;
+    }
+
+    // (4) Show all todos
+    else if (command === "What is on my todo?") {
+        return todos.length > 0 ? `You have ${todos.length} todos: ${todos.join(", ")}` : "Your todo list is empty.";
+    }
+
+    // (5) Remove item from the todo
+    else if (command.startsWith("Remove")) {
+        let todoItem = command.replace("Remove", "").replace("from my todo", "").trim();
+        let index = todos.indexOf(todoItem);
+        if (index !== -1) {
+            todos.splice(index, 1);
+            return `Removed ${todoItem} from your todo.`;
+        } else {
+            return `${todoItem} is not in your todo list`;
+        }
+    }
+
+    // (6) Get today's date
+    else if (command === "What day is it today?") {
+        let today = new Date();
+        let day = today.getDate();
+        let month = today.toLocaleString("en-US", { month: "long" });
+        let year = today.getFullYear();
+        return `${day}. of ${month} ${year}`;
+    }
+
+    // (7) Math
+    else if (command.startsWith("What is")) {
+        let expression = command.replace("What is", "").trim();
+        if (/^[0-9+\-*/().\s]+$/.test(expression)) {
+            try {
+                let result = eval(expression);
+                return `${expression} = ${result}`;
+            } catch (error) {
+                return "Sorry, I can't calculate that.";
+            }
+        } else {
+            return "Sorry, I can't calculate that.";
+        }
+    }
```

Comment:
Suggestion: [Return early](https://forum.freecodecamp.org/t/the-return-early-pattern-explained-with-javascript-examples/19364) pattern.

- Check for validation errors first and return early if they are found
- This way you avoid nesting ifs, which has a high cognitive load for humans

Reply:
Regexes can be very powerful, also pretty hard to debug.
Luckily, there are some [nice](https://regex-vis.com/) [tools](https://regex101.com/) out there.